### ISSUE TITLE
Multiple completion suggesters

### ISSUE THREAD

**Issue by redserpent7** [elastic/elasticsearch-net#1408]
Hi,

I am trying to create some completion suggesters on some of my fields. My document class looks like this:
```
[ElasticType(Name = "rawfiles", IdProperty = "guid")]
public class RAW
{
	[ElasticProperty(OmitNorms = true, Index = FieldIndexOption.NotAnalyzed, Type = FieldType.String, Store = true)]
    public string guid { get; set; }
			
	[ElasticProperty(OmitNorms = true, Index = FieldIndexOption.Analyzed, Type = FieldType.String, Store = true, IndexAnalyzer = "def_analyzer", SearchAnalyzer = "def_analyzer_search", AddSortField = true)]
	public string filename { get; set; }

	[ElasticProperty(OmitNorms = true, Index = FieldIndexOption.Analyzed, Type = FieldType.String, Store = true, IndexAnalyzer = "def_analyzer", SearchAnalyzer = "def_analyzer_search")]
    public List<string> tags { get { return new List<string>(); } }
}
```

And here is how I am trying to create the completion fields

```
public bool CreateMapping(ElasticClient client, string indexName)
{
    IIndicesResponse result = null;
    try
    {
        result = client.Map<RAW>(
                c => c.Index(indexName)
                    .MapFromAttributes()
                    .AllField(f => f.Enabled(false))
                    .SourceField(s => s.Enabled())
                    .Properties(p => p
                        .Completion(s => s.Name(n => n.tags.Suffix("comp"))
                                    .IndexAnalyzer("standard")
                                    .SearchAnalyzer("standard")
                                    .MaxInputLength(20)
                                    .Payloads()
                                    .PreservePositionIncrements()
                                    .PreserveSeparators())
                        .Completion(s2 => s2.Name(n=>n.filename.Suffix("comp"))
                                    .IndexAnalyzer("standard")
                                    .SearchAnalyzer("standard")
                                    .MaxInputLength(20)
                                    .Payloads()
                                    .PreservePositionIncrements()
                                    .PreserveSeparators())
                        )
                   );
    }
    catch (Exception)
    {

    }

    return result != null && result.Acknowledged;
}
```

My problem is that this is only creating a single completion field named "comp". I was under the impression that this will create two completion fields, one named filename.comp and the other named tags.comp.

I then tried the answer on this [SO question](http://stackoverflow.com/questions/21433586/nest-suggestcompletion-usage-throws-is-not-a-completion-suggest-field-excepti) but this complicated the matter even worse as now my two fields were mapped as a completion field only.

Just to be clear, I want to create a multi-field (field) that has a data, sort and completion fileds. Much like the one in [this example](https://developer.rackspace.com/blog/qbox/)

**Comment by gmarz** [comment]
Hey @redserpent7 apologies for the slow reply here.  I noticed you asked this on SO: http://stackoverflow.com/questions/30285065/elasticsearch-nest-client-creating-multi-field-fields-with-completion

Did that answer help clear things up?

Another observation I've made is that the article you linked (and consequently the answer on SO) doesn't make use of the completion suggester, as you're trying to do.

You've also uncovered a bug/missing feature here in NEST.  Technically, completion fields are supported as a multi field, although the functionality is limited (no access to inputs, output, payload, and weight).  However, NEST at the moment doesn't allow for this.  We'll fix this asap.

That said, in order to properly setup a completion type and gain access to all of the features, you should use the the baked in [SuggestField](https://github.com/elastic/elasticsearch-net/blob/develop/src/Nest/Domain/Suggest/SuggestField.cs), (or roll your own type if you need multiple inputs due to #1220).  This would require you to forego the multi field approach and add this type as another field(s) to your POCO.  As also mentioned in the SO answer, [check this out](https://github.com/elastic/elasticsearch-net-example#part-5-suggestions) for a full example.

Hope that helps.

**Comment by Mpdreamz** [comment]
Elasticsearch mapping fields using dots is a tad buggy see this issue: https://github.com/elastic/elasticsearch/issues/11337

I think the fix here is to map `tags` and `filename` separately either as an object with `Object => Properties => CompletionField` or as multified string with a completion sub field `String => Fields => CompletionField` depending on which best represent your POCO. 

ping @gmarz @gmoskovicz

**Comment by gmarz** [comment]
> Elasticsearch mapping fields using dots is a tad buggy see this issue: elastic/elasticsearch#11337

Yikes, nice find.

> I think the fix here is to map tags and filename separately and then the comp field either as an object with Object => Properties => CompletionField or as multified string with a completion sub field String => Fields => CompletionField depending on which best represent your POCO.

Agreed, but right now a multi field can't be mapped as a sub field: `Fields` takes a [CorePropertiesDescriptor](https://github.com/elastic/elasticsearch-net/blob/develop/src/Nest/Domain/Mapping/Descriptors/MultiFieldMappingDescriptor.cs#L28) which doesn't contain a `MultiFieldDescriptor`.

A simple fix could be overloading `Fields()` with a multi field version.

**Comment by Mpdreamz** [comment]
> A simple fix could be overloading Fields() with a multi field version.

:+1: to this but it might be a tad tricky since now we have a clear separation between CoreTypes and non core types

I think fields needs to take a [PropertiesDescriptor](https://github.com/elastic/elasticsearch-net/blob/1d996bf0862bd9dfaaef736fe2751cc0d9d81b74/src/Nest/Domain/Mapping/Descriptors/PropertiesDescriptor.cs) instead of [CorePropertiesDescriptor](https://github.com/elastic/elasticsearch-net/blob/1d996bf0862bd9dfaaef736fe2751cc0d9d81b74/src/Nest/Domain/Mapping/Descriptors/CorePropertiesDescriptor.cs) but that might not be possible in a `1.x` release.

**Comment by redserpent7** [comment]
Sorry for the late reply @gmarz but I did exactly as @Mpdreamz suggested in his comment.
I created a separate field for each suggester. First I created a class named `SuggestField`

```
public class SuggestField
        {
            public SuggestField()
            {
                Input = new List<string>();
            }

            public IEnumerable<string> Input { get; set; }
        }
```

Then I mapped two new fields in my document:

```
public SuggestField filenamesuggest 
            {
                get
                {
                    return new SuggestField { Input = new List<string> { filename }};
                } 
            }

public SuggestField tagssuggest
            {
                get
                {
                    return new SuggestField { Input = new List<string>(tags)};
                }
            }
```

And finally I mapped these fields as such:

```
public override bool CreateMapping(ElasticClient client, string indexName)
{
        IIndicesResponse result = null;
        try
        {
            result = client.Map<RAW>(
                    c => c.Index(indexName)
                        .MapFromAttributes()
                        .TimestampField(t => t.Enabled().Path(p => p.timestamp))
                        .AllField(f => f.Enabled(false))
                        .SourceField(s => s.Enabled())
                        .Properties(p => p
                            .Completion(s => s
                                .Name(n => n.filenamesuggest)
                                .IndexAnalyzer("suggest_analyzer")
                                .SearchAnalyzer("def_analyzer_search")
                                .MaxInputLength(20)
                                .Payloads()
                                .PreservePositionIncrements()
                                .PreserveSeparators())
                            .Completion(s => s
                                .Name(n=>n.tagssuggest)
                                .IndexAnalyzer("suggest_analyzer")
                                .SearchAnalyzer("def_analyzer_search")
                                .MaxInputLength(20)
                                .Payloads()
                                .PreservePositionIncrements()
                                .PreserveSeparators())
                            )
                       );
        }
        catch (Exception)
        {

        }

        return result != null && result.Acknowledged;
}
```

**Comment by redserpent7** [comment]
Ok so don't mind me asking this here but given my code above, how can I add a context to my suggester with NEST?

**Comment by gmarz** [comment]
@redserpent7 check out [this test](https://github.com/elastic/elasticsearch-net/blob/develop/src/Tests/Nest.Tests.Unit/Core/Map/Properties/PropertiesTests.cs#L358) for an example.  Let me know if that helps.

**Comment by gmarz** [comment]
This is fixed in 2.0 with the new mapping refactoring.
